src/librustc_mir/transform/acs_propagate.rs

```diff
@@ -0,0 +1,314 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This is Alias-Constant-Simplify propagation pass. This is a composition of three distinct
+//! dataflow passes: alias-propagation, constant-propagation and terminator simplification.
+//!
+//! All these are very similar in their nature:
+//!
+//!                 | Constant  |  Alias   | Simplify  |
+//!|----------------|-----------|----------|-----------|
+//!| Lattice Domain | Lvalue    | Lvalue   | Lvalue    |
+//!| Lattice Value  | Constant  | Lvalue   | Constant  |
+//!| Transfer       | x = const | x = lval | x = const |
+//!| Rewrite        | x → const | x → lval | T(x) → T' |
+//!| Bottom         | {}        | {}       | {}        |
+//!
+//! For all of them we will be using a lattice of Hashmap from Lvalue to
+//! WTop<Either<Lvalue, Constant>>
+//!
+//! My personal belief is that it should be possible to make a way to compose two hashmap lattices
+//! into one, but I can’t seem to get it just right yet, so we do the composing and decomposing
+//! manually here.
+
+use rustc_data_structures::fnv::FnvHashMap;
+use rustc::mir::repr::*;
+use rustc::mir::visit::{MutVisitor, LvalueContext};
+use rustc::mir::transform::lattice::{Lattice, WBottom};
+use rustc::mir::transform::dataflow::*;
+use rustc::mir::transform::{Pass, MirPass, MirSource};
+use rustc::ty::TyCtxt;
+use rustc::middle::const_val::ConstVal;
+use pretty;
+
+#[derive(PartialEq, Debug, Eq, Clone)]
+enum Either<'tcx> {
+    Lvalue(Lvalue<'tcx>),
+    Const(Constant<'tcx>),
+}
+
+#[derive(Debug, Clone)]
+struct AcsLattice<'tcx> {
+    known_values: FnvHashMap<Lvalue<'tcx>, Either<'tcx>>
+}
+
+impl<'tcx> Lattice for AcsLattice<'tcx> {
+    fn bottom() -> Self { unimplemented!() }
+    fn join(&mut self, other: &Self) -> bool {
+        let mut to_remove = vec![];
+
+        for (k, v) in &self.known_values {
+            if other.known_values.get(k).map_or(true, |other_v| other_v != v) {
+                to_remove.push(k.clone());
+            }
+        }
+
+        for k in &to_remove {
+            self.known_values.remove(k);
+        }
+
+        !to_remove.is_empty()
+    }
+}
+
+pub struct AcsPropagate;
+
+impl Pass for AcsPropagate {}
+
+impl<'tcx> MirPass<'tcx> for AcsPropagate {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>, src: MirSource, mir: &mut Mir<'tcx>) {
+        let ret = ar_forward(
+            &mut mir.cfg,
+            Facts::new(),
+            AcsPropagateTransfer,
+            AliasRewrite.and_then(ConstRewrite).and_then(SimplifyRewrite)
+        );
+        mir.cfg = ret.0;
+        pretty::dump_mir(tcx, "acs_propagate", &0, src, mir, None);
+    }
+
+}
+
+struct AcsPropagateTransfer;
+
+fn base_lvalue<'a, 'tcx>(mut lval: &'a Lvalue<'tcx>) -> &'a Lvalue<'tcx> {
```

Comment:
It'd be reasonable to just add this as a method on `Lvalue` since other transformations will almost certainly find it useful as well.